source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,598 @@
+           GetIntegerLiteral(type.GetInOperand(1u)) == 1)) {
+        std::vector<ir::Instruction*> decorations =
+            get_decoration_mgr()->GetDecorationsFor(type.result_id(), true);
+        if (decorations.empty()) {
+          id = type.result_id();
+          break;
+        }
+      }
+    }
+  }
+  if (id != 0) return id;
+
+  id = TakeNextId();
+  context()->AddType(MakeUnique<ir::Instruction>(
+      context(), SpvOpTypeInt, 0, id,
+      std::initializer_list<ir::Operand>{
+          {SPV_OPERAND_TYPE_LITERAL_INTEGER, {32}},
+          {SPV_OPERAND_TYPE_LITERAL_INTEGER, {0}}}));
+  ir::Instruction* integer = &*--context()->types_values_end();
+  get_def_use_mgr()->AnalyzeInstDef(integer);
+
+  return id;
+}
+
+void ScalarReplacementPass::GetOrCreateInitialValue(ir::Instruction* source,
+                                                    uint32_t index,
+                                                    ir::Instruction* newVar) {
+  assert(source->opcode() == SpvOpVariable);
+  if (source->NumInOperands() < 2) return;
+
+  uint32_t initId = source->GetSingleWordInOperand(1u);
+  uint32_t storageId = GetStorageType(newVar)->result_id();
+  ir::Instruction* init = get_def_use_mgr()->GetDef(initId);
+  uint32_t newInitId = 0;
+  if (init->opcode() == SpvOpConstantNull) {
+    // Initialize to appropriate NULL.
+    auto iter = type_to_null_.find(storageId);
+    if (iter == type_to_null_.end()) {
+      newInitId = TakeNextId();
+      type_to_null_[storageId] = newInitId;
+      context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+          context(), SpvOpConstantNull, storageId, newInitId,
+          std::initializer_list<ir::Operand>{}));
+      ir::Instruction* newNull = &*--context()->types_values_end();
+      get_def_use_mgr()->AnalyzeInstDefUse(newNull);
+    } else {
+      newInitId = iter->second;
+    }
+  } else if (ir::IsSpecConstantInst(init->opcode())) {
+    // Generate a constant for the extraction.
+    uint32_t constantId = 0;
+    auto iter = index_to_constant_.find(index);
+    if (iter == index_to_constant_.end()) {
+      if (int_id_ == 0) {
+        int_id_ = GetIntId();
+      }
+      constantId = TakeNextId();
+      context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+          context(), SpvOpConstant, int_id_, constantId,
+          std::initializer_list<ir::Operand>{
+              {SPV_OPERAND_TYPE_LITERAL_INTEGER, {index}}}));
+      ir::Instruction* constant = &*--context()->types_values_end();
+      get_def_use_mgr()->AnalyzeInstDefUse(constant);
+    } else {
+      constantId = iter->second;
+    }
+
+    // Create a new constant extract.
+    newInitId = TakeNextId();
+    context()->AddGlobalValue(MakeUnique<ir::Instruction>(
+        context(), SpvOpSpecConstantOp, storageId, newInitId,
+        std::initializer_list<ir::Operand>{
+            {SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER, {SpvOpCompositeExtract}},
+            {SPV_OPERAND_TYPE_ID, {constantId}}}));
```

Comment:
Are you sure this is the correct instruction?  An OpCompositeExtract takes more than 1 operand.

Reply:
I'll double check. It passes the validator though.

Reply:
You're right, I'll fix this.